src/jsonrpc.rs

```diff
@@ -0,0 +1,199 @@
+use crate::cli::TrinConfig;
+use reqwest::blocking as reqwest;
+use serde_json;
+use std::fs;
+use std::io::{self, Read, Write};
+use std::net::{TcpListener, TcpStream};
+use std::os::unix;
+use threadpool::ThreadPool;
+
+pub fn launch_trin(trin_config: TrinConfig, infura_project_id: String) {
+    let pool = ThreadPool::new(trin_config.pool_size as usize);
+
+    match trin_config.protocol.as_str() {
+        "ipc" => launch_ipc_client(pool, infura_project_id, &trin_config.ipc_path),
+        "http" => launch_http_client(pool, infura_project_id, trin_config),
+        val => panic!("Unsupported protocol: {}", val),
+    }
+}
+
+fn launch_ipc_client(pool: ThreadPool, infura_project_id: String, ipc_path: &String) {
+    let listener_result = unix::net::UnixListener::bind(ipc_path);
+    let listener = match listener_result {
+        Ok(listener) => listener,
+        Err(err) if err.kind() == io::ErrorKind::AddrInUse => {
+            // TODO something smarter than just dropping the existing file and/or
+            // make sure file gets cleaned up on shutdown.
+            match fs::remove_file(ipc_path) {
+                Err(_) => panic!("Could not serve IPC from existing path '{}'", ipc_path),
+                Ok(()) => unix::net::UnixListener::bind(ipc_path).unwrap(),
```

Comment:
So this approach no longer seems acceptable with a configurable IPC path (just dropping whatever file is found there and replacing it).

The simple answer is that we should just panic instead of trying to remove the file and reattempting. But then to solve the original problem: I think that we were leaking the file on shutdown sometimes. (Maybe anytime there's a panic?)

So the next step is to make sure that the IPC file gets removed when trin shuts down, even after a panic. (Although, it's probably worth verifying the theory. All I really know is that sometimes it used to happen that I had a pre-existing file on startup, and it was annoying. Not sure where it came from).

Reply:
I updated the code to add some handlers that will clean up the ipc file upon `ctrl+c` shutdown and all panics. It works as expected locally, but could use some integration tests to make sure the cleanup happens as expected. I'm leaning towards saving those for a larger integration test pr coming soon<sup>tm</sup>.